go/feast/featurestore.go

```diff
@@ -0,0 +1,907 @@
+	}
+
+	onDemandFeatureViews, err := fs.registry.listOnDemandFeatureViews(fs.config.Project)
+	if err != nil {
+		// TODO (Ly): Review: return an empty list instead of
+		// error if no OnDemandFeatureView is found in this project
+		onDemandFeatureViews = make([]*OnDemandFeatureView, 0)
+	}
+	for _, onDemandFeatureView := range onDemandFeatureViews {
+		odFvs[onDemandFeatureView.base.name] = onDemandFeatureView
+	}
+
+	if featureService, ok := parsedKind.(*FeatureService); ok {
+
+		fvsToUse := make([]*FeatureView, 0)
+		requestFvsToUse := make([]*RequestFeatureView, 0)
+		odFvsToUse := make([]*OnDemandFeatureView, 0)
+
+		for _, featureProjection := range featureService.projections {
+			// Create copies of FeatureView that may
+			// contains the same *FeatureView but
+			// each differentiated by a *FeatureViewProjection
+			featureViewName := featureProjection.name
+			if fv, ok := fvs[featureViewName]; ok {
+				base, err := fv.base.withProjection(featureProjection)
+				if err != nil {
+					return nil, nil, nil, err
+				}
+				fvsToUse = append(fvsToUse, fv.NewFeatureViewFromBase(base))
+			} else if requestFv, ok := requestFvs[featureViewName]; ok {
+				base, err := requestFv.base.withProjection(featureProjection)
+				if err != nil {
+					return nil, nil, nil, err
+				}
+				requestFvsToUse = append(requestFvsToUse, requestFv.NewRequestFeatureViewFromBase(base))
+			} else if odFv, ok := odFvs[featureViewName]; ok {
+				base, err := odFv.base.withProjection(featureProjection)
+				if err != nil {
+					return nil, nil, nil, err
+				}
+				odFvsToUse = append(odFvsToUse, odFv.NewOnDemandFeatureViewFromBase(base))
+			} else {
+				return nil, nil, nil, errors.New(fmt.Sprintf("The provided feature service %s contains a reference to a feature view"+
+					"%s which doesn't exist. Please make sure that you have created the feature view"+
+					"%s and that you have registered it by running \"apply\".", featureService.name, featureViewName, featureViewName))
+			}
+		}
+		return fvsToUse, requestFvsToUse, odFvsToUse, nil
+	}
+
+	fvsToUse := make([]*FeatureView, len(fvs))
+	requestFvsToUse := make([]*RequestFeatureView, len(requestFvs))
+	odFvsToUse := make([]*OnDemandFeatureView, len(odFvs))
+	index := 0
+	for _, fv := range fvs {
+		fvsToUse[index] = fv
+		index += 1
+	}
+
+	index = 0
+	for _, fv := range requestFvs {
+		requestFvsToUse[index] = fv
+		index += 1
+	}
+	index = 0
+	for _, fv := range odFvs {
+		odFvsToUse[index] = fv
+		index += 1
+	}
+	return fvsToUse, requestFvsToUse, odFvsToUse, nil
+}
+
+func (fs *FeatureStore) getEntityMaps(requestedFeatureViews []*FeatureView) (map[string]string, error) {
+
+	entityNameToJoinKeyMap := make(map[string]string)
+	var entityNames map[string]bool
+	var entityName string
+	var joinKeyMap map[string]string
+	var featureView *FeatureView
+
+	entities, err := fs.listEntities(true, false)
+
+	if err != nil {
+		entities = make([]*Entity, 0)
```

Comment:
What would happen here if the registry is an API and it's down. Don't we need to handle the error?